### PR TITLE
WIP: Via way restrictions

### PR DESCRIPTION
**Pull request by lukasalexanderweber**
background: https://discuss.graphhopper.com/t/add-support-for-complex-turn-restrictions/7326 

I implemented the reading of via way and multi via way turn restrictions. The next step would be to consider this information in the graph creation. For now, I would just ask you if you are fine with the way those turn restrictions are read?

### CODE REVIEW

**Comment by lukasalexanderweber** [comment]
At the moment I think that the Restriction graph approach of [OSRM](https://github.com/Project-OSRM/osrm-backend/pull/5907) could be applicable for graphhoppers graph creation, too. If you think so too, i would try to implement this.

File: core/src/main/java/com/graphhopper/reader/osm/OSMReader.java [hunk 1]
```diff
@@ -596,21 +599,45 @@ static List<OSMTurnRestriction> createTurnRelations(ReaderRelation relation, Str
...
+            // ONE FROM Way (a "no_entry" restriction can have ONE OR MORE FROM ways)
+            if (toWayID >= 0 && viaIDs.size() > 0) {
                 List<OSMTurnRestriction> res = new ArrayList<>(2);
                 for (ReaderRelation.Member member : relation.getMembers()) {
                     if (ReaderElement.Type.WAY == member.getType() && "from".equals(member.getRole())) {
-                        OSMTurnRestriction osmTurnRelation = new OSMTurnRestriction(member.getRef(), viaNodeID, toWayID, type);
+                        OSMTurnRestriction osmTurnRelation = new OSMTurnRestriction(member.getRef(), viaIDs, toWayID, type, viaType);
                         osmTurnRelation.setVehicleTypeRestricted(vehicleTypeRestricted);
                         osmTurnRelation.setVehicleTypesExcept(vehicleTypesExcept);
                         res.add(osmTurnRelation);
```
Comment (lukasalexanderweber) [inline core/src/main/java/com/graphhopper/reader/osm/OSMReader.java:643]: @karussell in https://github.com/graphhopper/graphhopper/pull/2648 you implemented this functionality. 
1. Is it intentional to restrict the possible OSMTurnRestriction List to 2 elements?
2. Should we add a check that the restriction is a "no_entry"?
3. Should we extend this approach to TO ways since there multiple ways for "no_exit" are possible as well?

Comment (lukasalexanderweber) [reply core/src/main/java/com/graphhopper/reader/osm/OSMReader.java:643]: > Do you mean the constructor argument of ArrayList? It is not restricted to 2 elements but just reduces the default from 10 but probably not necessary.

A I didn't know that its possible to specify the initialCapacity. I thought it limits the array 👍

File: core/src/main/java/com/graphhopper/reader/osm/OSMTurnRestriction.java [hunk 2]
```diff
@@ -24,40 +24,50 @@
...
+		return viaOSMIds;
+	}
+    
+    public RestrictionType getRestriction() {
         return restriction;
     }
+    
+	public ViaType getViaType() {
+		return viaType;
+	}
```
[the remark this answers is not in the document]
Comment (lukasalexanderweber) [reply core/src/main/java/com/graphhopper/reader/osm/OSMTurnRestriction.java:70]: When I use your code formatting xml in eclipse, it somehow changes lots of other pieces of code which I don't manipulate. Thats why I think it might be better if one of you does the formatting once after the functionality is implemented.

File: core/src/main/java/com/graphhopper/reader/osm/OSMTurnRestriction.java [hunk 3]
```diff
@@ -87,15 +97,10 @@ public boolean isVehicleTypeConcernedByTurnRestriction(Collection<String> vehicl
         return vehicleTypeRestricted.isEmpty() || vehicleTypes.contains(vehicleTypeRestricted);
     }
 
-    @Override
-    public String toString() {
-        return "*-(" + fromOsmWayId + ")->" + viaOsmNodeId + "-(" + toOsmWayId + ")->*";
-    }
-
-    public enum Type {
+	public enum RestrictionType {
```
[the remark this answers is not in the document]
Comment (lukasalexanderweber) [reply core/src/main/java/com/graphhopper/reader/osm/OSMTurnRestriction.java:100]: We now have two enums in the OSMTurnRestriction:

RestrictionType (if a restriction is Prohibitory (`no`) or Mandatory (`only`))
ViaType (if a restriction has a `via node`, a `via way` or `multi via ways`)

I thougt `RestrictionType and ViaType` is a better combination as `Type and ViaType`, but I am happy to reverse these changes :)

[the remark this answers is not in the document]
Comment (lukasalexanderweber) [reply core/src/main/java/com/graphhopper/reader/osm/OSMTurnRestriction.java:100]: what about keeping Type as is and adding Via?

```
    public enum Via {
        UNSUPPORTED, NODE, WAY, MULTI_WAY;
    }
```

File: core/src/main/java/com/graphhopper/reader/osm/OSMTurnRestriction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukasalexanderweber) [reply core/src/main/java/com/graphhopper/reader/osm/OSMTurnRestriction.java:70]: I'll check if I'm allowed to install community IntelliJ for at least the formatting

[the remark this answers is not in the document]
Comment (otbutz) [reply core/src/main/java/com/graphhopper/reader/osm/OSMTurnRestriction.java:70]: IANAL but Jetbrains is pretty clear about its usage terms: [Can I use Community Editions of JetBrains IDEs for developing commercial proprietary software?](https://sales.jetbrains.com/hc/en-gb/articles/360021922640-Can-I-use-Community-Editions-of-JetBrains-IDEs-for-developing-commercial-proprietary-software-)

tl;dr you should be safe even if you use it in a business environment

[the remark this answers is not in the document]
Comment (lukasalexanderweber) [reply core/src/main/java/com/graphhopper/reader/osm/OSMTurnRestriction.java]: I formatted with IntelliJ, it should now be fine

**Comment by lukasalexanderweber** [comment]
I tested this approach by manipulating OSM Data

graph

Given is the `no_u_turn `restriction from 1 via 2 to 3.

We can support it by:

1. adding a via-node restriction (1b2)
2. copying the edges and node of this via-node restriction (1b2)
3. adding a via-node restriction from the copied via-way edge (2n) to the to-way edge (2n c 3)

This forces the routing to go on the copied edge if it wants to go left and from there isn't able to turn left again.

---

The following needs to be considered: 

1. Removal of the copied node bn (as a shortcut) should be prevented, since then the navigation instruction to turn left would be missing [1].
2. Till now, unintentional navigation instructions are added (keep left or keep right), since the road technically splits into two ways. This needs to be prevented.
3. Snapping onto copied ways should be prevented

[1] this could be prevented by also copying all other edges leading to b, but I'm not sure if there might be a better option

the side effects 2 and 3 are quite similar to https://github.com/Project-OSRM/osrm-backend/issues/2681#issuecomment-311972554

---

an `only_u_turn` restriction could be supported with this approach as well:
1. blocking [2] the b node coming from 1, diverting onto the duplicated node and edges.
2. add a only restriction to (2n c 3)

[2] blocking could be done adding both a no and a only restriction (or implemented as so)

---

@karussell @easbar could you give feedback if you think this is feasable?

**Comment by lukasalexanderweber** [comment]
> See also [the existing issue](https://github.com/graphhopper/graphhopper/issues/446#issuecomment-507221777)...
> 
> >

thanks for this, I will have a look 👍 

> Maybe focus on the single via ways for now (especially when the multi-via-way handling is more complicated) and when everything is working (including the snap prevention etc) we revisit this :)
> 
> >

Yeah I think so, too. 

> What do you mean here?

If the node is copied with only the from- and to-edges, there will be no "turn left" instruction:

image

There will be one if another edge is added to the node:

image

(see also the unwanted "keep left" due to the artificial edge)

**Comment by lukasalexanderweber** [comment]
graph2

However, this will not be a problem with the approach of https://github.com/graphhopper/graphhopper/issues/446 (which I overlooked -.-)

We would:

1. we introduce an artificial node bn and edges a -> bn (1) and bn -> b (1n)
2. we add an artificial edge (oneway) bn -> c (2n)
3. we deny the turn 1n -> b -> 2
4. we deny the turn 2n -> c -> 3 
5. we deny the turn 3 -> c -> 2n

will the graph still be correct if the distance between bn and b is 0? So is there "place" for 1n?

**Comment by lukasalexanderweber** [comment]
Ok, to summarize:

Given the graph

original_s

with the `no_u_turn` restriction `from 1 via 2 to 3`

During graph preperation, we will:

1. introduce an artificial node `bn` and edges a -> bn `1n` and bn -> b `1`
2. add an artificial edge (oneway) bn -> c `2n`
3. deny the turn 1 -> b -> 2
4. deny the turn 2n -> c -> 3
5. deny the turn 1 -> bn -> 2n

umleitung_s

Notes:

- The distance between `b` and `bn` is 0
- The edge `2n` should be 'hidden' for the location index (see ferries: [#322](https://github.com/graphhopper/graphhopper/issues/322))

### Proof of Concept:

<details><summary>original.osm</summary>

```
<?xml version='1.0' encoding='UTF-8'?>
<osm version="0.6" generator="pbf2osm">
    <!-- 
	
	    10
	    |
    1~~~2~~~3~~~4
        |   |   |
        |   5~~~6 
        |   |
    7~~~8~~~9
        |
		11

    (1,2)->(2,8)->(8,7): no_u_turn
    (2,3)->(3,5)->(5,9)->(9,8): no_u_turn
    -->

    <node id="1" lat="0.050" lon="0.010" version="1"></node>
    <node id="2" lat="0.052" lon="0.010" version="1"></node>
    <node id="3" lat="0.054" lon="0.010" version="1"></node>
    <node id="4" lat="0.056" lon="0.010" version="1"></node>
    <node id="5" lat="0.054" lon="0.012" version="1"></node>
    <node id="6" lat="0.056" lon="0.012" version="1"></node>
    <node id="7" lat="0.050" lon="0.014" version="1"></node>
    <node id="8" lat="0.052" lon="0.014" version="1"></node>
    <node id="9" lat="0.054" lon="0.014" version="1"></node>
    <node id="10" lat="0.052" lon="0.008" version="1"></node>
    <node id="11" lat="0.052" lon="0.016" version="1"></node>

    <!-- 10-2 -->
    <way id="102" version="1">
        <nd ref="10"/>
        <nd ref="2"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 1-2 -->
    <way id="12" version="1">
        <nd ref="1"/>
        <nd ref="2"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 2-3 -->
    <way id="23" version="1">
        <nd ref="2"/>
        <nd ref="3"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 3-4 -->
    <way id="34" version="1">
        <nd ref="3"/>
        <nd ref="4"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 2-8 -->
    <way id="28" version="1">
        <nd ref="2"/>
        <nd ref="8"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 3-5 -->
    <way id="35" version="1">
        <nd ref="3"/>
        <nd ref="5"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 4-6 -->
    <way id="46" version="1">
        <nd ref="4"/>
        <nd ref="6"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 5-6 -->
    <way id="56" version="1">
        <nd ref="5"/>
        <nd ref="6"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 5-9 -->
    <way id="59" version="1">
        <nd ref="5"/>
        <nd ref="9"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 7-8 -->
    <way id="78" version="1">
        <nd ref="7"/>
        <nd ref="8"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 8-9 -->
    <way id="89" version="1">
        <nd ref="8"/>
        <nd ref="9"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 8-11 -->
    <way id="811" version="1">
        <nd ref="8"/>
        <nd ref="11"/>
        <tag k="highway" v="motorway"/>
    </way>

    <relation id="1001" version="1">
        <member type="way" ref="12" role="from"/>
        <member type="way" ref="28" role="via"/>
        <member type="way" ref="78" role="to"/>
        <tag k="restriction" v="no_u_turn"/>
        <tag k="type" v="restriction"/>
    </relation>

    <relation id="1002" version="1">
        <member type="way" ref="23" role="from"/>
        <member type="way" ref="35" role="via"/>
        <member type="way" ref="59" role="via"/>
        <member type="way" ref="89" role="to"/>
        <tag k="restriction" v="no_u_turn"/>
        <tag k="type" v="restriction"/>
    </relation>

</osm>
```

</details>

<details><summary>modified.osm</summary>

```
<?xml version='1.0' encoding='UTF-8'?>
<osm version="0.6" generator="pbf2osm">
    <!-- 
	
        10
        |
    1~~~2~~~3~~~4
        |   |   |
        |   5~~~6 
        |   |
    7~~~8~~~9
        |
        11

    (1,2)->(2,8)->(8,7): no_u_turn
    (2,3)->(3,5)->(5,9)->(9,8): no_u_turn
    -->

    <node id="1" lat="0.050" lon="0.010" version="1"></node>
    <node id="2" lat="0.052" lon="0.010" version="1"></node>
    <node id="3" lat="0.054" lon="0.010" version="1"></node>
    <node id="4" lat="0.056" lon="0.010" version="1"></node>
    <node id="5" lat="0.054" lon="0.012" version="1"></node>
    <node id="6" lat="0.056" lon="0.012" version="1"></node>
    <node id="7" lat="0.050" lon="0.014" version="1"></node>
    <node id="8" lat="0.052" lon="0.014" version="1"></node>
    <node id="80" lat="0.052" lon="0.014" version="1"></node>
    <node id="9" lat="0.054" lon="0.014" version="1"></node>
    <node id="10" lat="0.052" lon="0.008" version="1"></node>
    <node id="11" lat="0.052" lon="0.016" version="1"></node>

    <!-- 10-2 -->
    <way id="102" version="1">
        <nd ref="10"/>
        <nd ref="2"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 1-2 -->
    <way id="12" version="1">
        <nd ref="1"/>
        <nd ref="2"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 2-3 -->
    <way id="23" version="1">
        <nd ref="2"/>
        <nd ref="3"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 3-4 -->
    <way id="34" version="1">
        <nd ref="3"/>
        <nd ref="4"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 2-8 -->
    <way id="28" version="1">
        <nd ref="2"/>
        <nd ref="8"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 2-80 -->
    <way id="280" version="1">
        <nd ref="2"/>
        <nd ref="80"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 3-5 -->
    <way id="35" version="1">
        <nd ref="3"/>
        <nd ref="5"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 4-6 -->
    <way id="46" version="1">
        <nd ref="4"/>
        <nd ref="6"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 5-6 -->
    <way id="56" version="1">
        <nd ref="5"/>
        <nd ref="6"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 5-9 -->
    <way id="59" version="1">
        <nd ref="5"/>
        <nd ref="9"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 7-8n -->
    <way id="78" version="1">
        <nd ref="7"/>
        <nd ref="80"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 8-8n -->
    <way id="88" version="1">
        <nd ref="80"/>
        <nd ref="8"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 8-9 -->
    <way id="89" version="1">
        <nd ref="8"/>
        <nd ref="9"/>
        <tag k="highway" v="motorway"/>
    </way>

    <!-- 8-11 -->
    <way id="811" version="1">
        <nd ref="8"/>
        <nd ref="11"/>
        <tag k="highway" v="motorway"/>
    </way>

    <relation id="1001" version="1">
        <member type="way" ref="88" role="from"/>
        <member type="node" ref="8" role="via"/>
        <member type="way" ref="28" role="to"/>
        <tag k="restriction" v="no_left_turn"/>
        <tag k="type" v="restriction"/>
    </relation>

    <relation id="1002" version="1">
        <member type="way" ref="280" role="from"/>
        <member type="node" ref="2" role="via"/>
        <member type="way" ref="12" role="to"/>
        <tag k="restriction" v="no_left_turn"/>
        <tag k="type" v="restriction"/>
    </relation>

</osm>
```

</details>

Routing on `modified.osm` yields the desired result, preventing the u turn while allowing a left turn:

image

**Comment by lukasalexanderweber** [comment]
I might need some inspiration on where to manipulate the graph creation...

In OSMReader#preprocessRelations I could store all via-way restrictions. 

I then have the `from`, `to` and `via` way OSM Id of the Restrictions. 

Now I need the Node which links the  `from` and `via` ways (b in the example above). Any idea how to retrieve this node?

**Review by lukasalexanderweber — COMMENTED** [review]
...

**Review by lukasalexanderweber — COMMENTED** [review]
first tests are successfull for real life turn restrictions. Now some refinements are needed.

File: core/src/main/java/com/graphhopper/reader/osm/OSMTurnRestriction.java [hunk 4]
```diff
@@ -24,25 +24,37 @@
...
         this.toOsmWayId = toWayID;
         this.restriction = restrictionType;
+        this.viaType = viaType;
         this.vehicleTypeRestricted = "";
         this.vehicleTypesExcept = new ArrayList<>();
     }
 
+    public OSMTurnRestriction(long id, long fromWayID, long viaId, long toWayID, RestrictionType restrictionType, ViaType viaType) {
+        this(id, fromWayID, new ArrayList<>(Arrays.asList(viaId)), toWayID, restrictionType, viaType);
+    }
```
[the remark this answers is not in the document]
Comment (lukasalexanderweber) [reply core/src/main/java/com/graphhopper/reader/osm/OSMTurnRestriction.java:52]: yes during testing

File: core/src/main/java/com/graphhopper/reader/osm/OSMReader.java [hunk 5]
```diff
@@ -196,6 +209,94 @@ protected boolean acceptWay(ReaderWay way) {
...
+                }
+                try {
+                    NodeRestriction r = restriction.getRestrictions().get(0);
+                    NodeRestriction r2 = restriction.getRestrictions().get(1);
+    
+                    createArtificialViaNode(r.getVia(), nodeData);
+    
+                    // manipulate the first edge
+                    // 1.
+                    int from = nodeData.idToTowerNode(nodeData.getId(restriction.getStartNode()));
```
[the remark this answers is not in the document]
Comment (lukasalexanderweber) [reply core/src/main/java/com/graphhopper/reader/osm/OSMReader.java:234]: generally assert that all ways have exactly two Tower Nodes

File: core/src/main/java/com/graphhopper/reader/osm/OSMReader.java [hunk 6]
```diff
@@ -196,6 +209,94 @@ protected boolean acceptWay(ReaderWay way) {
...
+                    // 2.
+                    int from2 = nodeData.idToTowerNode(nodeData.getId(artificialViaNodes.get(r.getVia())));
+                    int to2 = nodeData.idToTowerNode(nodeData.getId(r.getVia()));
+                    PointList pointList2 = new PointList(2, nodeData.is3D());
+                    long newOsmId2 = nextArtificialOSMWayId++;
+                    osmWayIdSet.add(newOsmId2);
+                    LongArrayList longalist = new LongArrayList();
+                    longalist.add(artificialViaNodes.get(r.getVia()));
+                    longalist.add(r.getVia());
+                    ReaderWay artificial_way2 = new ReaderWay(newOsmId2, way.getTags(), longalist);
```
[the remark this answers is not in the document]
Comment (lukasalexanderweber) [reply core/src/main/java/com/graphhopper/reader/osm/OSMReader.java]: Maybe because of negative Ids?

File: core/src/main/java/com/graphhopper/reader/osm/OSMReader.java [hunk 7]
```diff
@@ -196,6 +209,94 @@ protected boolean acceptWay(ReaderWay way) {
...
+        }
+    }
+
+    protected void buildRestrictions(OSMNodeData nodeData) {
+        for (WayRestriction restriction : way_restrictions) {
+            // for now we only work with single via-way restrictions 
+            if (restriction.getWays().size() == 3) {
+                restriction.buildRestriction(osmWayNodesMap);
+                if (!restriction.isValid()) {
+                	continue;
```
[the remark this answers is not in the document]
Comment (lukasalexanderweber) [reply core/src/main/java/com/graphhopper/reader/osm/OSMReader.java]: Fallback

File: core/src/main/java/com/graphhopper/reader/osm/OSMReader.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (lukasalexanderweber) [reply core/src/main/java/com/graphhopper/reader/osm/OSMReader.java]: Further analysis shows that the navigation instructions are only manipulated on the edges with distance 0. Maybe we can collect some ideas on how to handle this.

**Comment by otbutz** [comment]
Could we simplify/breakup the OSMReader in a different PR?

**Comment by lukasalexanderweber** [comment]
yeah as things got bigger than expected, I will split this in multiple PRs.
